src/testing/main_HypSmearingTest.cu

```diff
@@ -0,0 +1,137 @@
+/* 
+ * main_HypSmearingTest.cu                                                               
+ * 
+ */
+
+#include "../SIMULATeQCD.h"
+#include "../modules/hyp/hypSmearing.h"
+
+#define PREC double
+#define MY_BLOCKSIZE 256
+#define USE_GPU true
+
+int main(int argc, char *argv[]) {
+
+    stdLogger.setVerbosity(INFO);
+    LatticeParameters param;
+    StopWatch<true> timer;
+    CommunicationBase commBase(&argc, &argv);
+    param.readfile(commBase, "../parameter/tests/hypSmearingTest.param", argc, argv);
+    commBase.init(param.nodeDim());
+
+    const size_t HaloDepth = 0;
+
+    rootLogger.info("Initialize Lattice");
+    typedef GIndexer<All,HaloDepth> GInd;
+    initIndexer(HaloDepth,param,commBase);
+
+    rootLogger.info("Initialize Gaugefield");
+    Gaugefield<PREC, true, HaloDepth> gauge_in(commBase);
+    //Gaugefield<PREC, true, HaloDepth> gauge_Lvl1(commBase);
+    //Gaugefield<PREC, false,HaloDepth> gauge_Lvl1_host(commBase);
+    //Gaugefield<PREC, false,HaloDepth> gauge_smeared_bielefeldgpu(commBase);
+    //Gaugefield<PREC, true, HaloDepth> gauge_u3(commBase);
+    //Gaugefield<PREC, false,HaloDepth> gauge_u3_host(commBase);
+    //Gaugefield<PREC, true, HaloDepth> gauge_Lv2(commBase);
+    //Gaugefield<PREC, false,HaloDepth> gauge_Lv2_host(commBase);
+    Gaugefield<PREC, true, HaloDepth> gauge_out(commBase);
+    
+    timer.start();
+    //HypSmearing<PREC, USE_GPU, HaloDepth,R18> smearing(gauge_in, gauge_out);
+    HypSmearing<PREC, USE_GPU, HaloDepth,R18> smearing(gauge_in);
+    timer.stop();
+    rootLogger.info("Time for initializing HypSmearing: ", timer);
+    timer.reset();
+
+    rootLogger.info("Read configuration");
+    gauge_in.readconf_nersc("../test_conf/lat.sample.l6666_nersc");
```

Comment:
Please replace this readconf line with a dynamically generated config from "GaugeUpdate::set_gauge_to_reference()" and remove lat.sample.l6666_nersc from the branch

Reply:
Hi Lucas,

I based the Hyp smearing code on the Hisq smearing code, which still has:
gauge_in.readconf_nersc("../test_conf/gauge12750");
in main_HisqSmearingTest.cpp
I would request clarification on if this code is also going to be modified to use the dynamic scheme or not.  I think there should be one policy for both tests, in principle.

Also, I validated the hyp smearing code based on the lattice included.  It seems this is a request that I output the dynamically generated lattice, smear it with the Milc code, and then check that the output is the same.  If this is not necessary, than it seems we are saying that the barrier to the merge is not the correctness of the smearing.  If that is the case, can you yourself be the one to make the change after the merge?  I'm not much interested in the semantics of the test (or which lattice is used).    

Is the dynamic lattice the same from run to run (what if the RNG generator used changes, or the implementation changes how it handles the zero seed)?  The way the test works now is string comparison vs. the outputs of Milc code.  I don't currently have another way to validate it.  If you do have a dynamic validation strategy, though, I would be interested to know what it is and if it can be added as an additional (or substitute) check.  

Cheers,

Reply:
Hi Lukas,

Confirming the RNG does not depend on implementation (i.e. if I compile it on my laptop it will give the same output as on a KNL) addresses one of my concerns in my above comment.  The other concern I had is that main_HisqSmearingTest.cpp is not using a dynamic lattice.  If this is the policy, then you should make the change there too.  It is slightly beyond the scope of this discussion, but I had a look again at main_HisqSmearingTest.cpp.  The test used is a unitarity test, so it should not be lattice dependent.  However, main_HisqSmearingTest.cpp uses a reference lattice + unitarity test (which is lattice independent), and I don't know how this reference lattice was generated.  If, as you indicate, the policy involved is to accept the correctness of the test as is, then it should be straightforward to generate this reference lattice dynamically.  

My other concern (as stated) is that I'm overall not interested in what lattice is used, so it does not make much sense for me to be further involved in making this change.  You are welcome to pull my changes and make this change afterwards; it is clear that it is not critical in the physics sense whether the change is made before or after (if it were, then main_HisqSmearingTest would've already been changed).  

I can confirm that I did the test with the Milc code as you describe.

Hopefully this resolves this issue; I'm uninterested in repeating myself.

Reply:
Dear Dan,

I agree with you that main_HisqSmearingTest could also be updated. This is not surprising, since this test is rather old and the policy that we should use dynamic lattices is rather new. 
An issue to replace the usage of static lattices with dynamic lattices in all tests already exists. 
And someone should definitely take care of that at some point.
However, the fact that main_HisqSmearingTest is not using a dynamic lattice, does not justify that all new contributions are allowed to use configurations from disk.

I really appreciate that you want to make a contribution to the code and I am always happy to help if you face any problems in doing so. 
However, I will not finalize this branch for you.  
If you don't want to do that, you will probably need to find someone else to do it for you. 
Otherwise, we cannot merge this branch. These are the rules.

Regards,
Lukas